browser-hist.el

```diff
@@ -83,55 +98,60 @@ db, we copy the file."
          (new-fname (format "%sbhist-%s.sqlite"
                             (temporary-file-directory)
                             (symbol-name browser))))
-    (copy-file hist-db new-fname :overwite)
-    new-fname))
-
-(defun browser-hist--query (browser)
-  "Query db."
-  (let* ((built-in? (and (fboundp 'sqlite-available-p)
-                         (sqlite-available-p)))
-         (db-file (browser-hist--make-db-copy browser))
-         (query (alist-get browser browser-hist--db-queries))
-         (db (if built-in?
-                 (sqlite-open db-file)
-               (progn
-                 (require 'sqlite)
-                 (sqlite-init db-file))))
-         (rows
-          (thread-last
-            (if built-in?
-                (sqlite-select db query)
-             (sqlite-query db query))
-            (seq-remove
-             (lambda (x) (or (null (car x)) (string-blank-p (car x)))))
-            (seq-map
-             (lambda (x)
-               (when (and (car x) (cadr x))
-                (cons (string-trim-right
-                       (replace-regexp-in-string
-                        (if browser-hist-ignore-query-params "\\?.*" "")
-                        "" (cadr x)) "/")
-                      (car x))))))))
+    (if (or (not (file-exists-p new-fname))
+            (file-newer-than-file-p hist-db new-fname))
+        (copy-file hist-db new-fname :overwite :keep-time)
+      new-fname)))
+
+(defvar browser-hist--db-connection nil)
+
+(defun browser-hist--send-query (strings)
+  "Build the sql query from the minibuffer input"
+  (pcase-let* ((`(,title ,url ,table ,rest)
+                (alist-get browser-hist-default-browser
+                           browser-hist--db-fields))
+               (full-query
+                (cl-loop for s in (split-string strings)
+                         collect (format " ( %s LIKE '%%%s%%' OR %s LIKE '%%%s%%' ) "
+                                         title s url s)
+                         into queries
+                         finally return
+                         (concat (format "SELECT DISTINCT %s, %s FROM %s WHERE"
+                                  title url table)
+                                 (mapconcat #'identity queries " AND ")
+                                 rest)))
```

Comment:
Wow, crafting beautiful sushi with cl-loop. As someone who never learned the magic and intricacies of cl-loop, I have great respect for people who know how to use it. Very nice.

Reply:
Thanks, I just threw that one in without much thought. `cl-loop`s are usually better than `map` in performance-critical "hot-loops", but it's probably not required here.

That said, the updated version reduces the whole function to two cl-loops, including handling the case of no input (where it returns the latest history).